timewarp_gl/plugin.cpp

```diff
@@ -185,10 +311,75 @@ class timewarp_gl : public threadloop {
+            return GL_SRGB8_ALPHA8;
+        default:
+            return 0;
+        }
+    }
+
+    void ImportVulkanImage(const vk_image_handle& vk_handle, swapchain_usage usage) {
+        [[maybe_unused]] const bool gl_result = static_cast<bool>(glXMakeCurrent(dpy, root, glc));
+        assert(gl_result && "glXMakeCurrent should not fail");
+        assert(GLEW_EXT_memory_object_fd && "[timewarp_gl] Missing object memory extensions for Vulkan-GL interop");
+
+        // first get the memory handle of the vulkan object
+        GLuint memory_handle;
+        GLint  dedicated = GL_TRUE;
+        glCreateMemoryObjectsEXT(1, &memory_handle);
+        glMemoryObjectParameterivEXT(memory_handle, GL_DEDICATED_MEMORY_OBJECT_EXT, &dedicated);
+        glImportMemoryFdEXT(memory_handle, vk_handle.allocation_size, GL_HANDLE_TYPE_OPAQUE_FD_EXT, vk_handle.file_descriptor);
+
+        // then use the imported memory as the opengl texture.
+        // since we're writing to an intermediate texture that's the same memory format as Monado's layer renderer,
+        // there's no need to reformat anything.
+        GLuint format = ConvertVkFormatToGL(vk_handle.format);
+        assert(format != 0 && "Given Vulkan format not handled!");
+        GLuint image_handle;
+        glGenTextures(1, &image_handle);
+        glBindTexture(GL_TEXTURE_2D, image_handle);
+        glTextureStorageMem2DEXT(image_handle, 1, format, vk_handle.width, vk_handle.height, memory_handle, 0);
+
+        float color[4] = {0.0f, 0.0f, 0.0f, 1.0f};
+        glTexParameterfv(GL_TEXTURE_2D, GL_TEXTURE_BORDER_COLOR, color);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_BORDER);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_BORDER);
+
+        switch (usage) {
+        case swapchain_usage::LEFT_SWAPCHAIN: {
+            _m_eye_swapchains[0].push_back(image_handle);
+            break;
+        }
+        case swapchain_usage::RIGHT_SWAPCHAIN: {
+            _m_eye_swapchains[1].push_back(image_handle);
+            break;
+        }
+        case swapchain_usage::LEFT_RENDER: {
+            _m_eye_output_textures[0] = image_handle;
+            break;
+        }
+        case swapchain_usage::RIGHT_RENDER: {
+            _m_eye_output_textures[1] = image_handle;
+            break;
+        }
+        default: {
+            assert(false && "Invalid swapchain usage");
+            break;
+        }
+        }
+    }
+
     void BuildTimewarp(HMD::hmd_info_t& hmdInfo) {
         // Calculate the number of vertices+indices in the distortion mesh.
         num_distortion_vertices = (hmdInfo.eyeTilesHigh + 1) * (hmdInfo.eyeTilesWide + 1);
         num_distortion_indices  = hmdInfo.eyeTilesHigh * hmdInfo.eyeTilesWide * 6;
+        // What is the 6 here? What's the difference between vertices and indices? Each rectangular has two triangles, each
+        // triangle has three vertices (i.e., three indices to represent).
```

Comment:
Seems like leftover comments for understanding the code. Remove if not no longer needed.